src/document.cpp

```diff
@@ -58,199 +59,234 @@ using namespace std::placeholders;  // _1, _2, etc for std::bind
 static const char* const unsupported = "Internal error: Feature not supported in this mode.";
 static const char* const invalid_type = "$Invalid_type";
 
-string location_t::str() const
+template <typename T, typename = void>
+struct has_print_indent : std::false_type
+{};
+
+template <typename T>
+struct has_print_indent<
+    T, std::void_t<decltype(std::declval<T&>().print(std::declval<std::ostream&>, std::declval<const std::string&>))>>
+    : std::true_type
+{};
+
+template <typename T>
+constexpr auto has_print_indent_v = has_print_indent<T>::value;
```

Comment:
This seems excessive for the few cases that have an indent parameter. Can't we just call `T.print(os, INDENT)` and avoid the meta programming

Reply:
With concepts it's going to be much much simpler, I promise :)
No, we cannot just call `print(os, INDENT)` because the `chan_priority_t` does not have it.

Reply:
Obviously we wouldnt call `chan_priority_t::print(os, INDENT)`, I was thinking of the few outliers that actually use an indent parameter. But looking through the code I dont see a single print function that uses it right now. It might be useful later.